### PR TITLE
[Feat] Enable external serving of Vizro assets

### PR DESCRIPTION
**Pull request by antonymilne**
## Description

* Change our weird (but remarkably effective) flask blueprint scheme for serving Vizro assets into a more conventional Dash library scheme for handling resources 
* This makes it possible to serve all our resources through the CDN when `serve_locally=False` (previously this was possible only for the library CSS, which was just `figures.css`)
* Fix serving of fonts, which previously was putting extraneous `<link rel="stylesheet">` in our HTML source
* Update JS so that it works in-browser without being a module - this is needed due to the new way assets are served

To do:
- [x] @antonymilne: maybe write unit tests
- [x] @antonymilne: check https://vizro--775.org.readthedocs.build/en/775/pages/user-guides/run/
- [x] @huong-li-nguyen: manual test + make sure you understand how everything works with the following - I've already done the first two but good to have a second pair of eyes on it. Please check the 404 page on these and also that the order of CSS is ok since I didn't check that.
  - [x] `serve_locally=True` Vizro app (try it disconnected from Internet)
  - [x] `serve_locally=False` Vizro app
  - [x] `serve_locally=True` Dash app using Vizro library
  - [x] `serve_locally=False` Dash app using Vizro library (less important to test, because it's an edge case and also if the above work then I'm confident this does too)

## Notes (will go in [updated developer docs](https://github.com/mckinsey/vizro/wiki/Developer-docs))

When `serve_locally=True` (the default), Dash serves component library resources (generally CSS and JS) through the Flask server using the `_dash-component-suites` route.
* For Vizro library components (currently just KPI cards), this should happen _when Vizro is imported_.
* For Vizro framework components (everything else), this should happen _only when `Vizro()` is instantiated_. 

This makes our footprint as small as possible and ensures there's reduced risk of CSS name clash when someone wants to use Vizro just as a library but doesn't instantiate `Vizro()` (not common at all now, but maybe will be in the future).

When `serve_locally=False`, Dash switches to serving resources through `external_url`, where it's specified. For Vizro components we use jsDeliver as a CDN for this.

A few complications:
* files that aren't CSS/JS (e.g. fonts, maps) can still be served through the same routes but should not have a `<script>` or `<link>` in the HTML source. This is achieved with `dynamic=True`
* the CDN minifies CSS/JS files automatically, but some we have minified manually ourselves (currently just `vizro-bootstrap.min.css`)
* it's not possible to serve JS as modules this way, which means we can't easily do `import/export` in them 

In future:
* when we release vizro-bootstrap, nothing changes for Vizro framework. Pure Dash users would treat it like any other bootstrap theme i.e. set it through `external_stylesheets` that points to the stylesheet on our CDN or download the file to their own `assets` folder. We'd have a shortcut for this like `vizro.theme` or do it through bootswatch if that's possible.
* ideally we would webpack our JS and ship the `.min.js` rather than just relying on the CDN to minify it. This would let us write "proper" rather than just in-browser JS and mean we benefit from tree-shaking etc. rather than just minification that the CDN does. In reality the optimisations would make very little difference to performance, but it's kind of the "right" way to do things. It's more effort than it's worth to set up at the moment, but if we end up maintaining a bigger JS codebase we might do it
* vizro-boostrap.css map file + all the SASS should live in the repo so that it can be handled correctly through developer tools in a browser

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/src/vizro/_vizro.py [hunk 1]
```diff
@@ -173,13 +163,41 @@ def _reset():
...
-        we will get ["vizro/css/accordion.css", "vizro/css/button.css", ...].
-        """
-        return sorted(
-            (STATIC_URL_PREFIX / path.relative_to(folder)).as_posix() for path in folder.rglob(f"*.{extension}")
+def _make_resource_spec(path: Path):
+    # For dev versions, a branch or tag called e.g. 0.1.20.dev0 does not exist and so won't work with the CDN. We point
+    # to main instead, but this can be manually overridden to the current feature branch name if required.
+
+    # _git_branch = __version__ if "dev" not in __version__ else "main"
+    _git_branch = "feat/allow-servering-external-assets"
```
Comment (antonymilne) [inline vizro-core/src/vizro/_vizro.py:172]:
```suggestion
    _git_branch = __version__ if "dev" not in __version__ else "main"
```
Must be applied before merging

File: vizro-core/src/vizro/static/js/actions/build_action_loop_callbacks.js [hunk 2]
```diff
@@ -76,7 +76,16 @@ export function _gateway(
   return [remaining_actions.slice(1)].concat(trigger_next);
 }
 
-export function _after_action_cycle_breaker(data) {
+function after_action_cycle_breaker(data) {
   document.getElementById("cycle_breaker_div").click();
   return [];
 }
+
+window.dash_clientside = {
```
[the remark this answers is not in the document]
Comment (petar-qb) [reply vizro-core/src/vizro/static/js/actions/build_action_loop_callbacks.js:84]: Fantastic! Now, if you put the following code under the `assets/js/custom.js`, it will overwrite the Vizro `collapse_nav_panel` properly and will disable the nav_panel to collapse automatically:

```js
window.dash_clientside.dashboard.collapse_nav_panel = function (n_clicks, is_open) {
    if (!n_clicks) {
      return dash_clientside.no_update;
    }
    if (is_open) {
        return [
          false,
          {
            transform: "rotate(180deg)",
            transition: "transform 0.35s ease-in-out",
          },
          "Show Menu",
        ];
    } else {
      return [
        true,
        {
          transform: "rotate(0deg)",
          transition: "transform 0.35s ease-in-out",
        },
        "Hide Menu",
      ];
    }
}
```

File: vizro-core/src/vizro/_vizro.py [hunk 3]
```diff
@@ -46,41 +58,18 @@ def __init__(self, **kwargs):
...
+        for path in set(VIZRO_ASSETS_PATH.rglob("*")) - _library_css_files - _library_js_files:
+            if path.suffix == ".css":
+                self.dash.css.append_css(_make_resource_spec(path))
+            elif path.suffix == ".js":
+                self.dash.scripts.append_script(_make_resource_spec(path))
+            else:
+                # map files and fonts and images. These are treated like scripts since this is how Dash handles them.
+                # This adds paths to self.dash.registered_paths so that they can be accessed without throwing an
+                # error in dash._validate.validate_js_path.
+                self.dash.scripts.append_script(_make_resource_spec(path))
```
Comment (petar-qb) [inline vizro-core/src/vizro/_vizro.py:72]: I 👍  keeping this separate although the code block under the `elif` and `else` is the same.

File: vizro-core/src/vizro/_vizro.py [hunk 4]
```diff
@@ -46,41 +59,18 @@ def __init__(self, **kwargs):
...
-        # Exclude vizro/css/figures.css since these are distributed through the vizro._css_dist mechanism.
-        # In future we will probably handle all assets this way and none of this code will be required.
-        vizro_css = [
-            requests_pathname_prefix + path
-            for path in self._get_external_assets(vizro_assets_folder, "css")
-            if path != "vizro/css/figures.css"
-        ]
-
-        # Ensure vizro-bootstrap.min.css is loaded in first to allow overwrites
-        vizro_css.sort(key=lambda x: not x.endswith("vizro-bootstrap.min.css"))
```
[the remark this answers is not in the document]
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/_vizro.py:62]: This was mainly to ensure that if a user wants to overwrite any bootstrap class names, this still works 👍 I'll test it out manually later, but it's better to have this loaded in first.

File: vizro-core/src/vizro/static/css/vizro-bootstrap.min.css
[the remark this answers is not in the document]
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/static/css/vizro-bootstrap.min.css:1]: Ah, got it. I guess performance-wise it will be the same, as the minified version will be used in any case. Adding the unminified css file would just add lots of code to our repository, but I guess that's fine then?

**Review by petar-qb — COMMENTED** [review]
I guess these tickets could be closed after merging this: 
https://github.com/McK-Internal/vizro-internal/issues/1050 and 
https://github.com/McK-Internal/vizro-internal/issues/1267

**Comment by huong-li-nguyen** [comment]
> When `serve_locally=True`, Dash switches to serving resources through `external_url`, where it's specified. For Vizro components we use jsDeliver as a CDN for this.

I think you meant when `serve_locally=False` ?

**Comment by huong-li-nguyen** [comment]
> * [x]  @huong-li-nguyen: manual test + make sure you understand how everything works with the following - I've already done the first two but good to have a second pair of eyes on it. Please check the 404 page on these and also that the order of CSS is ok since I didn't check that.
>   
>   * [ ]  `serve_locally=True` Vizro app (try it disconnected from Internet)
>   * [ ]  `serve_locally=False` Vizro app
>   * [ ]  `serve_locally=True` Dash app using Vizro library
>   * [ ]  `serve_locally=False` Dash app using Vizro library (less important to test, because it's an edge case and also if the above work then I'm confident this does too)

I've tested all above manually and everything seems to work fine 🥳 Just the order of reading in the vizro-bootstrap.min.css file I would change to how it was before. It's safer so that we can ensure overwrites work as expected.


| Manual test case                                   | Expected Result                                                                                                                                                                                                                                  | ✅ |
|----------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---|
| `serve_locally=True`  Vizro app (w/o internet)     | - Looks as expected <br> - `<link rel=stylesheet>` for all our CSS/JS files via `href="/_dash_component-suites/vizro/..."` <br> - user assets via `href "/assets"`  <br> no `<link rel=stylesheet>` for fonts                                    | ✅ |
| `serve_locally=False` Vizro app                    | - Looks as expected <br> - `<link rel=stylesheet>` for all our CSS/JS files via `href="https://cdn.jsdelivr.net/gh/mckinsey..."` <br> - user assets via `href "/assets"`  <br> no `<link rel=stylesheet>` for fonts                              | ✅ |
| `serve_locally=True` Dash app using Vizro library  | - Looks as expected <br> - `<link rel=stylesheet>` only for `figures.css` via `href="/_dash_component-suites/vizro/..."` <br> - user assets via `href "/assets"` <br> - no `<link rel=stylesheet>` for fonts or any other of our css files | ✅  |
| `serve_locally=False` Dash app using Vizro library | - Looks as expected <br> - `<link rel=stylesheet>` only for `figures.css` via `href="https://cdn.jsdelivr.net/gh/mckinsey..." <br> - user assets via `href "/assets"` <br> - no `<link rel=stylesheet>` for fonts or any other of our css files  | ✅ |

**Review by huong-li-nguyen — APPROVED** [review]
Approval under the condition to change the order of reading in the` vizro-bootstrap.min.css` file (should be read in first) before `figures.css` or any other of our css files 🚀

File: vizro-core/docs/pages/user-guides/run.md [hunk 5]
```diff
@@ -147,4 +147,8 @@ A Vizro app wraps a Dash app, which itself wraps a Flask app. Hence to deploy a
...
 
-[`Vizro`][vizro.Vizro] accepts `**kwargs` that are passed through to `Dash`. This enables you to configure the underlying Dash app using the same [arguments that are available](https://dash.plotly.com/reference#dash.dash) in `Dash`. For example, in a deployment context, you might like to specify a custom `url_base_pathname` to serve your Vizro app at a specific URL rather than at your domain root.
+[`Vizro`][vizro.Vizro] accepts `**kwargs` that are passed through to `Dash`. This enables you to configure the underlying Dash app using the same [arguments that are available](https://dash.plotly.com/reference#dash.dash) in `Dash`. For example, in a deployment context, some of the below arguments may be useful:
+
+- `url_base_pathname`: serve your Vizro app at a specific path rather than at the domain root, e.g. if you host your dashboard at http://www.example.com/my_dashboard/ then you would set `url_base_pathname="/my_dashboard"` .
```
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/run.md:152]:
```suggestion
- `url_base_pathname`: serve your Vizro app at a specific path rather than at the domain root, e.g. if you host your dashboard at http://www.example.com/my_dashboard/ then you would set `url_base_pathname="/my_dashboard/"` .
```

File: vizro-core/src/vizro/_vizro.py (hunk 3 above)
Comment (antonymilne) [reply vizro-core/src/vizro/_vizro.py:72]: Haha yes, I spent a while thinking about which I preferred and decided this would be clearest so glad you agree!

File: vizro-core/src/vizro/models/_dashboard.py [hunk 6]
```diff
@@ -303,9 +304,13 @@ def _make_page_layout(self, page: Page, **kwargs):
 
     @staticmethod
     def _make_page_404_layout():
+        # The svg file is available through the _dash-component-suites/vizro route, as used in Dash's
+        # _relative_url_path, but that feels too private to access directly. Hence read the file in directly rather
+        # than referring to its path.
+        error_404_svg = base64.b64encode((VIZRO_ASSETS_PATH / "images/error_404.svg").read_bytes()).decode("utf-8")
```
[the remark this answers is not in the document]
Comment (antonymilne) [reply vizro-core/src/vizro/models/_dashboard.py:310]: Basically never - we only do it this way because this is a special case where we have an image that's served from Vizro rather than the user assets folder. If we need to do this more in future then we should ideally have a better way of doing it because it's not very nice. But just for a one off it's ok.

File: vizro-core/src/vizro/_vizro.py (hunk 1 above)
Comment (antonymilne) [reply vizro-core/src/vizro/_vizro.py:171]: Ah yes, will do! The time delay is ok, it's just something to do with the CDN content caching I guess. If it causes problems with testing then it's possible to clear it manually: https://www.jsdelivr.com/tools/purge.

btw we should only very rarely need to set `_git_branch` manually here. It would really only be case where for some reason you need to test with `serve_locally=False` AND you've changed some of the assets. Apart from PRs like this that won't ever really happen.

File: vizro-core/src/vizro/_vizro.py [hunk 7]
```diff
@@ -173,13 +162,41 @@ def _reset():
...
-        """Returns a list of paths to assets with given `extension` in `folder`, prefixed with `STATIC_URL_PREFIX`.
 
-        e.g. with STATIC_URL_PREFIX="vizro", extension="css", folder="/path/to/vizro/vizro-core/src/vizro/static",
-        we will get ["vizro/css/accordion.css", "vizro/css/button.css", ...].
-        """
-        return sorted(
-            (STATIC_URL_PREFIX / path.relative_to(folder)).as_posix() for path in folder.rglob(f"*.{extension}")
+def _make_resource_spec(path: Path):
+    # For dev versions, a branch or tag called e.g. 0.1.20.dev0 does not exist and so won't work with the CDN. We point
+    # to main instead, but this can be manually overridden to the current feature branch name if required.
```
Comment (antonymilne) [inline vizro-core/src/vizro/_vizro.py:168]:
```suggestion
    # For dev versions, a branch or tag called e.g. 0.1.20.dev0 does not exist and so won't work with the CDN. We point
    # to main instead, but this can be manually overridden to the current feature branch name if required.
    # This would only be the case where you need to test something with serve_locally=False and have changed
    # assets compared to main. In this case you need to push your assets changes to remote for the CDN to update,
    # and it might also be necessary to clear the CDN cache: https://www.jsdelivr.com/tools/purge.
```

File: vizro-core/src/vizro/static/css/vizro-bootstrap.min.css
[the remark this answers is not in the document]
Comment (antonymilne) [reply vizro-core/src/vizro/static/css/vizro-bootstrap.min.css:1]: The minification is provided through the CDN so they minified version will only be used in the case you've set `serve_locally=False`. If you've just left this on the default (= 99.999% of users) you just get whatever is in our assets folder. So performance would be very slightly worse for these users, so ideally we would have the minified version here.

File: vizro-core/src/vizro/_vizro.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (antonymilne) [reply vizro-core/src/vizro/_vizro.py:62]: Got it, thank you! Made it so that `vizro-boostrap.min.css` is first in [ad943ab](https://github.com/mckinsey/vizro/pull/775/commits/ad943ab6807779652ddc367bdde7f47a49295046) and screenshots are now passing.

The full order is now:
1. vizro library assets in arbitrary order
2. vizro framework assets with `vizro-boostrap.min.css` first and rest in arbitrary order
3. user assets - not sure what order

Note that your example about `figures.css` won't actually work since as the moment library assets are always before the framework ones. We could probably change this if we needed to but since it doesn't cause any issues and might never do I'll leave it.

image

[the remark this answers is not in the document]
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/_vizro.py:62]: Alright, shall we read in things in alphabetical order (except for the bootstrap stylesheet)? This is what Dash does under the hood when reading the user assets folder

**Comment by petar-qb** [comment]
@antonymilne @huong-li-nguyen 

> Ah that would be a nice bonus! I don't understand these tickets well enough to check unfortunately. Please could you test when you review to see if they're fixed?

So, there are two tickets that I thought are solved with this PR:
 1. https://github.com/McK-Internal/vizro-internal/issues/1267 

The problem was that user's `js` scripts added to the `/assets` folder were loaded **before** Vizro `js` `/static` scripts. This disabled the possibility to overwrite vizro js default functions.

I guess that the `self.dash.scripts.append_script(...)` line solved this issue. (or moving `clientside_functions.js` logic to other files)

Here's an example on how it can be tested. 

Let's say that you want to disable automatically collapsing of the nav_panel when the screen's height/width is less than 576px (the behaviour that's implemented within the `dashboard.js` in the function called `collapse_nav_panel` that's assigned as a `dash.clientside_callback` in the `Dashboard.build`).  

From the `main` branch:
If you add the following custom `js` code inside the `assets/js/custom.js`, execution of this script will raise an exception: 
`"Cannot read properties of undefined (reading 'clientside')"` (this script is executed too early).
```js 
window.dash_clientside.clientside.collapse_nav_panel = function (n_clicks, is_open) {
  return [
    true,
    {
      transform: "rotate(0deg)",
      transition: "transform 0.35s ease-in-out",
    },
    "Hide Menu",
  ];
}
```

From the `feat/allow-servering-external-assets` branch:
if the same function is assigned to the `window.dash_clientside.dashboard.collapse_nav_panel` from the same file (assets/js/custom.js), this custom javascript function will properly overwrite default one (from the dashboard.js).


So, when this PR is merged, we can close the Issue -> https://github.com/McK-Internal/vizro-internal/issues/1267 

---

 2. https://github.com/McK-Internal/vizro-internal/issues/1050

FYI @nadijagraca 

This issue is not introduced with Vizro, and it exists in the dash app as well (for multi page apps). Also, this issue is not solved with this PR (and probably can't be solved because it's not the issue at all 😄). The issue says that if you, for example, add something like `document.getElementById("page-1-component-1-id").addEventListener("click", myScript);` directly to the `assets/js/custom.js`, this will be executed before the `page-1` is opened. As javascript can't find the "page-1-component-1-id" component, it means that `myScript` will not be executed when the component is clicked. All we can do is to explain to user different ways to overcome this problem (as it's already explained in the issue description)

File: vizro-core/changelog.d/20241002_113350_antony.milne_allow_servering_external_assets.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+### Deprecated
+
+- A bullet item for the Deprecated category with a link to the relevant PR at the end of your entry, e.g. Enable feature XXX ([#1](https://github.com/mckinsey/vizro/pull/1))
+
+-->
+
+### Fixed
+
+- Remove extraneous `<link>` to font file ([#775](https://github.com/mckinsey/vizro/pull/775))
+
```
[the remark this answers is not in the document]
Comment (antonymilne) [reply vizro-core/changelog.d/20241002_113350_antony.milne_allow_servering_external_assets.md:41]: Yes, done in [15dd8af](https://github.com/mckinsey/vizro/pull/775/commits/15dd8af24b1ccb4d4adbf57eb2129acebf022a3e).

File: vizro-core/docs/pages/user-guides/run.md [hunk 9]
```diff
@@ -147,4 +147,8 @@ A Vizro app wraps a Dash app, which itself wraps a Flask app. Hence to deploy a
 
 Internally, `app = Vizro()` contains a Flask app in `app.dash.server`. However, as a convenience, the Vizro `app` itself implements the [WSGI application interface](https://werkzeug.palletsprojects.com/en/3.0.x/terms/#wsgi) as a shortcut to the underlying Flask app. This means that, as in the [above example with Gunicorn](#gunicorn), the Vizro `app` object can be directly supplied to the WSGI server.
 
-[`Vizro`][vizro.Vizro] accepts `**kwargs` that are passed through to `Dash`. This enables you to configure the underlying Dash app using the same [arguments that are available](https://dash.plotly.com/reference#dash.dash) in `Dash`. For example, in a deployment context, you might like to specify a custom `url_base_pathname` to serve your Vizro app at a specific URL rather than at your domain root.
+[`Vizro`][vizro.Vizro] accepts `**kwargs` that are passed through to `Dash`. This enables you to configure the underlying Dash app using the same [arguments that are available](https://dash.plotly.com/reference#dash.dash) in `Dash`. For example, in a deployment context, some of the below arguments may be useful:
```
[the remark this answers is not in the document]
Comment (antonymilne) [reply vizro-core/docs/pages/user-guides/run.md:150]: I don't think so. These arguments are very obscure or there's no way anyone would try to override them. Given how many arguments are available in `Dash` I'd prefer to keep these docs simple and just emphasise the arguments that a user might actually want to alter.

File: vizro-core/src/vizro/_vizro.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (antonymilne) [reply vizro-core/src/vizro/_vizro.py:62]: Yes, good idea! Done in [efef57e](https://github.com/mckinsey/vizro/pull/775/commits/efef57e0fb29c26393041891cc94dab4987260d8).

Now the order is:

1. vizro library assets in alphabetical order
2. vizro framework assets with vizro-boostrap.min.css first and rest in alphabetical order
3. user assets in alphabetical order

**Comment by antonymilne** [comment]
Thanks very much for the careful reviewing @petar-qb and @huong-li-nguyen, much appreciated! I wrote some tests in [1ac4535](https://github.com/mckinsey/vizro/pull/775/commits/1ac4535b578608d7ea9e7f9fb883e3ff1fd73b06) and now it's merging :shipit:
